### DISCUSSION TITLE
Abort queued actions on button if variable is false

### DISCUSSION THREAD

**Discussion by Danie10** [bitfocus/companion#2973] [Q&A]
I'm trying to understand if what I'm trying to do is a bug or whether there is some other way to do it. I'm wanting do do the following:
1. Test if a window name is open and set focus to it - I run an external script that returns true/false and sets internal variable (works)
2. If internal variable is false (window not found open) the abort any further queued steps for this button (not working)
3. Otherwise, action 3 will output text sting or keys (works)

Action 1 and 3 work just fine, but I'm not getting action 2 to abort executing action 3. It always outputs the text to the current window. If the target window is actually open, it is changing focus to it and sending the text or keys to it. But I'd like to also have a failsafe that it only does so if the window is found open.

I ended up with the action 'internal: Actions: Abort delayed actions on a button' as being the most likely one to try for Action 2 above. I could have opted for some form of Trigger, but it would mean more scheduled triggers running in the background which is messy.

 So I have the action as it appears below here, and I set it to expression, and in the expression I put `$variable=='false'` but it is not aborting the queued Action 3. I have small delays set between the actions to allow them to fire one after each other.

![Screenshot_20244706_024717](https://github.com/user-attachments/assets/562523e5-16a7-40d4-8e56-10dd91fecc66)

So what I'm trying to find out is:
1. Is this not what the abort action actually does? In other words, it does not stop further actions on the same button?
2. The expression is somehow not valid or syntax is wrong? The internal variable is certainly updating to true/false.
3. There is better way to abort queued action based on a variable's value?

**Note:** I should just add this is on Linux, and when I'm in the X11 desktop of course this is not needed as xdotool will direct a key press or typed text straight at a named window all in one command. I'm trying to get this to work with Wayland, so I have to use the ydotool command, and that command cannot do such direction into a named window. The only way is to change focus to the window, and then output the keys as a second action step.

**Comment by Julusian** [comment]
I think you are hitting the problem that the actions don't wait for the previous one to complete before executing the next.
While it sometimes works with internal actions which are quick, an action which runs an external script will take some time to execute, and the following action won't wait for it to complete before running.

**Reply by Danie10** [reply]
Thanks, yes it was initially that before I forgot to add delays (was 0 and 0). But Now have 500ms and 500ms with relative delays and it's still doing it. I tried now lengthening to 1000ms each step as well. I have another external script that checks if some windows are open, and then changes the page, and that is running OK in fact on 200ms and 100ms.
